### ISSUE TITLE
HTTP GET challenge forwarded to backend

### ISSUE THREAD

**Issue by fabricetriboix** [JrCs/docker-letsencrypt-nginx-proxy-companion#223]
Hello,

I am trying to set up nginx:alpine + jwilder/docker-gen + jrcs/letsencrypt-nginx-proxy-companion. I want nginx to reverse proxy a wordpress website using HTTPS.

The problem I am having is that letsencrypt fails, apparently because it can't GET the challenge file. If I access the URL of the challenge file using a browser, the standard wordpress 404 page comes up. This means nginx just forwarded the request to the website being proxied instead of serving the challenge file.

Apparently, this problem should have been fixed by [this commit](https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion/pull/192/commits/c2cf98598018f6a680e30b19515d563c31986146), but apparently that's not enough for my setup...

Any idea?

Interesting part of the logs:

    nginx-letsencrypt    | Creating/renewal new.incise.co certificates... (new.incise.co)
    nginx-gen            | 2017/06/28 10:37:04 Contents of /etc/nginx/conf.d/default.conf did not change. Skipping notification ''
    nginx-letsencrypt    | 2017/06/28 10:37:04 Contents of /app/letsencrypt_service_data did not change. Skipping notification '/app/update_certs'
    nginx-letsencrypt    | 2017-06-28 10:37:05,268:INFO:simp_le:1211: Generating new account key
    nginx-letsencrypt    | 2017-06-28 10:37:06,856:INFO:requests.packages.urllib3.connectionpool:756: Starting new HTTPS connection (1): acme-v01.api.letsencrypt.org
    nginx-letsencrypt    | 2017-06-28 10:37:08,020:INFO:requests.packages.urllib3.connectionpool:756: Starting new HTTPS connection (1): letsencrypt.org
    nginx-letsencrypt    | 2017-06-28 10:37:09,139:INFO:requests.packages.urllib3.connectionpool:207: Starting new HTTP connection (1): new.incise.co
    nginx                | new.incise.co 172.18.0.1 - - [28/Jun/2017:10:37:09 +0000] "GET /.well-known/acme-challenge/rJjQu_MTPLcY2fnJBFzuVcle4Crs0hs5LhQ43vN89rI HTTP/1.1" 404 49331 "-" "python-requests/2.8.1"
    nginx                | 2017/06/28 10:37:09 [info] 6#6: *1 client 172.18.0.1 closed keepalive connection
    nginx-letsencrypt    | 2017-06-28 10:37:09,582:WARNING:simp_le:1303: new.incise.co was not successfully self-verified. CA is likely to fail as well!

Opening the [challenge URL](http://new.incise.co/.well-known/acme-challenge/rJjQu_MTPLcY2fnJBFzuVcle4Crs0hs5LhQ43vN89rI) with a browser opens the standard wordpress 404 page.

Here is my docker-compose file (edited to remove irrelevant stuff):

    version: "2"
    services:
     mysql:
      <irrelevant>
     nginx:
      image: nginx:alpine
      container_name: nginx
      restart: always
      ports:
       - "80:80"
       - "443:443"
      volumes:
       - /home/docker-stuff/certs/:/etc/nginx/certs/:ro
       - /home/docker-stuff/nginx/etc-nginx-conf.d/:/etc/nginx/conf.d/
       - /home/docker-stuff/nginx/etc-nginx-vhost.d/:/etc/nginx/vhost.d/
       - /home/docker-stuff/nginx/usr-share-nginx-html/:/usr/share/nginx/html/
       - /home/docker-stuff/nginx/nginx.conf:/etc/nginx/nginx.conf  # just to add debug logs
      labels:
       - "com.github.jrcs.letsencrypt_nginx_proxy_companion.nginx_proxy=true"
     docker-gen:
      image: jwilder/docker-gen:latest
      container_name: nginx-gen
      restart: always
      depends_on:
       - nginx
      volumes_from:
       - nginx
      volumes:
       - /home/docker-stuff/docker-gen/nginx.tmpl:/etc/docker-gen/templates/nginx.tmpl:ro
       - /var/run/docker.sock:/tmp/docker.sock:ro
      command: -notify-sighup nginx -watch -wait 5s:30s /etc/docker-gen/templates/nginx.tmpl /etc/nginx/conf.d/default.conf
     nginx-letsencrypt:
      image: jrcs/letsencrypt-nginx-proxy-companion:latest
      container_name: nginx-letsencrypt
      restart: always
      depends_on:
       - nginx
       - docker-gen
      environment:
       - NGINX_DOCKER_GEN_CONTAINER=nginx-gen
      volumes_from:
       - nginx
      volumes:
       - /home/docker-stuff/certs/:/etc/nginx/certs:rw
       - /var/run/docker.sock:/var/run/docker.sock:ro
     wordpress:
      environment:
       - VIRTUAL_HOST=new.incise.co
       - LETSENCRYPT_HOST=new.incise.co
       - LETSENCRYPT_EMAIL=contact@incise.co
      <rest is irrelevant for this topic>

Many thanks for any help!

**Comment by pLesur** [comment]
Hi, I'm having the same issue. I guess you weren't able to find a workaround @fabricetriboix  ?

**Comment by fabricetriboix** [comment]
Hi pLesur,

In the end, I had to get someone to look at it for me. It looks like the "vhost.d" stuff actually is a problem. Here is a diff of the nginx.tmpl file the guy fixed:

--- nginx.tmpl-orig	2017-07-14 15:58:02.658991594 +0200
+++ nginx.tmpl.orig	2017-07-21 16:16:48.000000000 +0200
@@ -207,9 +207,13 @@
 
 	{{ if (exists (printf "/etc/nginx/vhost.d/%s" $host)) }}
 	include {{ printf "/etc/nginx/vhost.d/%s" $host }};
-	{{ else if (exists "/etc/nginx/vhost.d/default") }}
-	include /etc/nginx/vhost.d/default;
-	{{ end }}
+#	{{ else if (exists "/etc/nginx/vhost.d/default") }}
+#	include /etc/nginx/vhost.d/default;
+#	{{ end }}
+	location /.well-known/acme-challenge {
+        	root /usr/share/nginx/html;
+	}
+
 
 	location / {
 		{{ if eq $proto "uwsgi" }}
@@ -246,9 +250,14 @@
 
 	{{ if (exists (printf "/etc/nginx/vhost.d/%s" $host)) }}
 	include {{ printf "/etc/nginx/vhost.d/%s" $host }};
-	{{ else if (exists "/etc/nginx/vhost.d/default") }}
-	include /etc/nginx/vhost.d/default;
-	{{ end }}
+#	{{ else if (exists "/etc/nginx/vhost.d/default") }}
+#	include /etc/nginx/vhost.d/default;
+#	{{ end }}
+
+	location /.well-known/acme-challenge {
+        	root /usr/share/nginx/html;
+	}
+
 
 	location / {
 		{{ if eq $proto "uwsgi" }}

In other words, do not include any "vhost" generated files, and tell nginx to serve the acme thing directly.

HTH!

 Fabrice

**Comment by davidbarratt** [comment]
I'm also having this problem... it looks like the `vhost.d` files are __never__ generated. I'm not really sure why they are not, but when I go into `/etc/nginx/vhost.d` they are missing. However, the image attempts to validate the cert, but it's forwarded to the backend. :(

**Comment by davidbarratt** [comment]
Ah ha!

in my case, both `dockergen` and `letsencrypt` needed access to the `vhost` volume.

**Comment by buchdag** [comment]
Yes, that's not explicitly stated in the doc but `docker-gen` needs access to `vhost` (and gets it in the provided example thanks to `--volumes-from nginx`).

**Comment by buchdag** [comment]
Closing due to inactivity and probably not being relevant anymore.
